### PR TITLE
ft.create vector field API update

### PR DESCRIPTION
**Pull request by GuyAv46**

### CODE REVIEW

**Review by DvirDukhan — CHANGES_REQUESTED** [review]
Looks clean and pretty 
Please see my comments about the error messages content
Also, docs need to be updated too

File: src/spec.c [hunk 1]
```diff
@@ -319,124 +319,241 @@ static int parseTextField(FieldSpec *fs, ArgsCursor *ac, QueryError *status) {
...
     return 0;
   }
-  if (!strncasecmp(VECSIM_ALGORITHM_BF, algStr, len)) 
-    algorithm = fs->vecSimParams.algo = VecSimAlgo_BF;
-  else if (!strncasecmp(VECSIM_ALGORITHM_HNSW, algStr, len)) 
-    algorithm = fs->vecSimParams.algo = VecSimAlgo_HNSWLIB;
-  else {
-    QERR_MKBADARGS_AC(status, "vecsim algorithm", AC_ERR_PARSE);
+  if (!mandtype) {
+    QERR_MKBADARGS_FMT(status, "Missing mandatory parameter %s", VECSIM_TYPE);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/spec.c:436]: Is "Missing mandatory parameter: cannot create %s index without specifying %s argument" better?

[the remark this answers is not in the document]
Comment (ashtul) [reply src/spec.c:436]:
```suggestion
    QERR_MKBADARGS_FMT(status, "missing %s for field %s", VECSIM_TYPE, fs->name);
```

**Review by ashtul — COMMENTED** [review]
Small suggestion.

File: tests/pytests/test_vecsim.py [hunk 2]
```diff
@@ -227,53 +227,74 @@ def testDelReuseLarge(env):
...
-    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLOAT32').error().contains('Bad arguments for vecsim size')
-    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLOAT32', '1024').error().contains('Bad arguments for vecsim metric')
-    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLOAT32', '1024', 'IP').error().contains('Bad arguments for vecsim algorithm')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR').error().contains('Bad arguments for vector similarity algorithm')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT').error().contains('Bad arguments for vector similarity number of parameters')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '6').error().contains('Expected 6 parameters but got 0')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '0').error().contains('Missing mandatory parameter')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '1').error().contains('expected even number')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '2', 'SIZE').error().contains('Bad arguments for algorithm')
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:261]: 'Bad arguments for algorithm'?
can this be more descriptive?

File: tests/pytests/test_vecsim.py [hunk 3]
```diff
@@ -227,53 +227,74 @@ def testDelReuseLarge(env):
...
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '2', 'TYPE').error().contains('Bad arguments for vector similarity FLAT index type')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '2', 'TYPE', 'FLOAT32').error().contains('Missing mandatory parameter')
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:263]: can you specify which parameter is missing?

File: tests/pytests/test_vecsim.py [hunk 4]
```diff
@@ -227,53 +227,74 @@ def testDelReuseLarge(env):
...
-    info = [['identifier', 'v', 'attribute', 'v', 'type', 'VECTOR', 'TYPE', 'FLOAT32', 'SIZE', '1024', 'METRIC', 'IP', 'ALGORITHM', 'HNSW', 'M', '16', 'EF CONSTRUCTION', '200']]
+    conn.execute_command('FT.CREATE', 'idx1', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '14', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC', 'IP', 'INITIAL_CAP', '10', 'M', '16', 'EF_CONSTRUCTION', '200', 'EF_RUNTIME', '10')
+    info = [['identifier', 'v', 'attribute', 'v', 'type', 'VECTOR', 'ALGORITHM', 'HNSW', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC', 'IP', 'M', '16', 'EF_CONSTRUCTION', '200', 'EF_RUNTIME', '10']]
     assertInfoField(env, 'idx1', 'attributes', info)
 
-    conn.execute_command('FT.CREATE', 'idx2', 'SCHEMA', 'v', 'VECTOR', 'FLOAT64', '4096', 'L2', 'HNSW', 'INITIAL_CAP', '10', 'M', '32', 'EF', '100')
-    info = [['identifier', 'v', 'attribute', 'v', 'type', 'VECTOR', 'TYPE', 'FLOAT64', 'SIZE', '4096', 'METRIC', 'L2', 'ALGORITHM', 'HNSW', 'M', '32', 'EF CONSTRUCTION', '100']]
-    assertInfoField(env, 'idx2', 'attributes', info)
+    # Uncomment these test when support for FLOAT64, INT32, INT64, is added.
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:234]: Those tests should cause failure for now. What will be the outcome if someone will try to run those commands?

File: tests/pytests/test_vecsim.py [hunk 5]
```diff
@@ -227,53 +227,74 @@ def testDelReuseLarge(env):
...
+        .error().contains('Bad arguments for vector similarity FLAT index blocksize')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '12', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC', 'IP', 'INITIAL_CAP', 'str', 'M', '16', 'EF_CONSTRUCTION', '200') \
+        .error().contains('Bad arguments for vector similarity HNSW index initial cap')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '12', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC', 'IP', 'INITIAL_CAP', '100', 'M', 'str', 'EF_CONSTRUCTION', '200') \
+        .error().contains('Bad arguments for vector similarity HNSW index m')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '12', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC', 'IP', 'INITIAL_CAP', '100', 'M', '16', 'EF_CONSTRUCTION', 'str') \
+        .error().contains('Bad arguments for vector similarity HNSW index ef')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '12', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC', 'IP', 'INITIAL_CAP', '100', 'M', '16', 'EF_RUNTIME', 'str') \
+        .error().contains('Bad arguments for vector similarity HNSW index ef')
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:294]: which `ef` parameter are we talking about here?

File: tests/pytests/test_vecsim.py [hunk 6]
```diff
@@ -227,53 +227,74 @@ def testDelReuseLarge(env):
...
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '2', 'TYPE', 'FLOAT32').error().contains('Missing mandatory parameter')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '4', 'TYPE', 'FLOAT32', 'DIM').error().contains('Bad arguments for vector similarity FLAT index dim')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '4', 'TYPE', 'FLOAT32', 'DIM', '1024').error().contains('Missing mandatory parameter')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'FLAT', '6', 'TYPE', 'FLOAT32', 'DIM', '1024', 'DISTANCE_METRIC').error().contains('Bad arguments for vector similarity FLAT index metric')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW').error().contains('Bad arguments for vector similarity number of parameters')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '6').error().contains('Expected 6 parameters but got 0')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '0').error().contains('Missing mandatory parameter')
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 'v', 'VECTOR', 'HNSW', '1').error().contains('expected even number')
```
Comment (DvirDukhan) [inline tests/pytests/test_vecsim.py:270]: can you write `expected an even number of parameters`?

File: tests/pytests/test_vecsim.py (hunk 4 above)
Comment (GuyAv46) [reply tests/pytests/test_vecsim.py:234]: We're testing for bad type valuse in line 279.
Right now these values will also result in the same error, but these tests will pass when we'll support them.
I can add tests for that in the 'invalid init args' section and comment to remove them when we add the right support.

**Review by DvirDukhan — APPROVED** [review]
LGTM
@ashtul

**Review by ashtul — APPROVED** [review]
Error messages can get some work but not a blocker

**Pull request merged**